### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by hbredin** [pyannote/pyannote-audio#220]
I think we should give/improve the following points : 
1) A high-level overview of the pipeline. How everything works, how each module is articulated with the others, etc ... 
2) More details about some mechanisms : I am thinking about user-defined callbacks since I've been working on that. But I'm pretty sure many of you will have other ideas 🙂
3) More explanations about some tricks that have been implemented : sliding window during the test, scheduler for the learning rate, data augmentation...
4) Some tips for users who might want to integrate their own code in pyannote (but that point goes well with the first one).

If each of us participate, we can write an extensive documentation that would potentially maximize users understanding of the pipeline, and would be a breeding ground for future pull requests.

I am aware that this does represent an extra workload, but if each of us does a tiny effort, I think we can greatly improve this aspect of pyannote.

I'd be glad to hear (read) your thoughts about that.

Marvin

_Originally posted by @MarvinLvn in https://github.com/pyannote/pyannote-audio/issues/219#issuecomment-535890214_

**Comment by hbredin** [comment]
I used sphinx and numpydoc for pyannote.metrics and pyannote.core.
I liked it because the whole documentation was also available as (either module or class) docstrings.  

"A high-level overview" would be the docstring of `pyannote.audio`.
"User-defined callbacks" would be the docstring of `pyannote.audio.train.callback`.
etc...

Then, it is just a matter of writting index files like [this one](https://github.com/pyannote/pyannote-metrics/blob/master/docs/source/index.rst) to glue them together.

Any experience in documentation packages @MarvinLvn @hadware?

**Comment by hadware** [comment]
Yes i've already used it on several occasions. The autodoc feature is pretty powerful. I don't think @MarvinLvn  already has, but I can teach him the basics (if he really needs some help, it's not like it's some kind of SuperVAD bash pipeline or anything).

**Comment by hbredin** [comment]
Work on documentation will take place in new branch `v2`.
In there, I would like to 
* add [type hints](https://docs.python.org/3/library/typing.html) all over the place
* automate documentation publishing to Github Pages with Github Actions

**Comment by hadware** [comment]
After the LREC deadline, I'll be up for type hintingization.

**Comment by hbredin** [comment]
FYI, I just setup a Github Action for automatic deployment of sphinx documentation.

**Comment by hbredin** [comment]
Here are a bunch of things that needs to be documented (list is not exhaustive):

  - Command line interface (`pyannote.audio.applications`)
    - Configuration file 
  - Training (`pyannote.audio.train`)
    - Base trainer (`pyannote.audio.train.trainer.Trainer`)
    - Task (`pyannote.audio.train.task.Task`)
    - Base model (`pyannote.audio.train.model.Model`)
    - Generators (`pyannote.audio.train.generator`)
      - Base batch generator (`pyannote.audio.train.generator.BatchGenerator`)
      - SmartBackground generator (`pyannote.audio.train.generator.SmartBackground`)
    - Callbacks (`pyannote.audio.train.callback.Callback`)
      - Logging (`pyannote.audio.train.logging.Logging`)
      - Schedulers (`pyannote.audio.train.schedulers`)
  - Feature extraction (`pyannote.audio.features`)
  - Data augmentation (`pyannote.audio.augmentation`)
  - Models (`pyannote.audio.models`)
  - Labeling (`pyannote.audio.labeling`)
    - Base labeling
    - Speech activity detection
    - Speaker change detection
    - Overlapped speech detection
    - Resegmentation
  - Embedding (`pyannote.audio.embedding`)
    - Base embedding
    - Triplet loss
    - Other losses 
  - Pipeline
  - Pretrained models
    - `torch.hub`
    - `pyannote.audio.labeling.extraction.SequenceLabeling`
    - `pyannote.audio.embedding.extraction.SequenceEmbedding`

**Comment by hbredin** [comment]
Would be great if the documentation was directly included in docstring and then imported in the actual sphinx documentation with autodoc.

**Comment by hbredin** [comment]
> Work on documentation will take place in new branch `v2`.

Branch v2 has been merged into develop. So work on documentation should be done as pull requests directly on develop.

**Comment by hbredin** [comment]
https://github.com/ines/course-starter-python could be an option.
